README.md
# KTBTaskQueue

[![Build Status](https://travis-ci.org/kevboh/KTBTaskQueue.png?branch=master)](https://travis-ci.org/kevboh/KTBTaskQueue)
[![Version](http://cocoapod-badges.herokuapp.com/v/KTBTaskQueue/badge.png)](http://cocoadocs.org/docsets/KTBTaskQueue)
[![Platform](http://cocoapod-badges.herokuapp.com/p/KTBTaskQueue/badge.png)](http://cocoadocs.org/docsets/KTBTaskQueue)

`KTBTaskQueue` is an optionally persistent queue that makes sure tasks get finished. It will prompt a delegate or block with a task until that task is declared successful or abandoned. Failed tasks can be retried immediately or at some point in the future, with the default following an exponential backoff pattern. The basic flow goes like this:

1. Put a task in the queue.
2. The queue hands the task to its delegate or an execution block. Whatever is executing the task... executes the task. You decide what executing a task looks like.
3. If the task is successful it's removed from the queue. If not, it's kept around and retried until it succeeds or it retries a `maxRetries` threshhold, after which it is abandonded and removed from the queue.

`KTBTaskQueue` can be used to track any task you wish. I find it useful to make sure network requests are successful when online and to keep them for later when offline.

## Using It

KTBTaskQueue is available through [CocoaPods](http://cocoapods.org), to install
it simply add the following line to your Podfile:

    pod "KTBTaskQueue"

Then in code, let’s say you use [AFNetworking](https://github.com/AFNetworking/AFNetworking) and want to make sure a particularly important POST reaches your server. You could do something like:

    // You created this queue in your app delegate or view controller or model.
    [self.queue enqueueTaskWithName:@"PostThisVitalThing" userInfo:@{"wow" : "so important"}];

Okay, now the queue will bug its delegate until the `PostThisVitalThing` task is complete. In the delegate:

    - (void)taskQueue:(KTBTaskQueue *)queue executeTask:(KTBTask *)task completion:(KTBTaskCompletionBlock)completion {
        // The task's userInfo is this request's parameters
        AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
        [manager POST:@"http://example.com/resources.json" parameters:task.userInfo success:^(AFHTTPRequestOperation *operation, id responseObject) {
            NSLog(@"JSON: %@", responseObject);
            completion(KTBTaskStatusSuccess);
        } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
            NSLog(@"Error: %@", error);
            completion(KTBTaskStatusFailure);
        }];
    }

By default, `PostThisVitalThing` will be retried after failure up to 10 times. That number can be customized, along with tons of other aspects of a task, by using a longer constructor available through `KTBTask`:

    [self.queue enqueueTask:[KTBTask taskWithName:@"FutureTask"
                                         userInfo:nil
                                    availableDate:[NSDate dateWithTimeIntervalSinceNow:60]
                                       maxRetries:3
                                       useBackoff:NO]];

`FutureTask` won't be attempted until a minute from now, will only be retried 3 times (for a total of four attempts), and will be retried immediately upon failure—the queue won't delay retry using its backoff technique.

The `KTBTaskQueueDelegate` offers more flexibility. It includes optional methods to report when tasks are abandoned and allows a custom delay to be applied in place of the built-in backoff delay. Check out `KTBTaskQueue.h` and `KTBTaskQueueDelegate.h` for more details and things to play with.

There's also an `executionBlock` property on `KTBTaskQueue` if you hate the delegate pattern:

    // Set the block that does the task
    queue.executionBlock = ^(KTBTask *task, KTBTaskCompletionBlock completion) {
        // The task's userInfo is this request's parameters
        AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
        [manager POST:@"http://example.com/resources.json" parameters:task.userInfo success:^(AFHTTPRequestOperation *operation, id responseObject) {
            NSLog(@"JSON: %@", responseObject);
            completion(KTBTaskStatusSuccess);
        } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
            NSLog(@"Error: %@", error);
            completion(KTBTaskStatusFailure);
        }];
    };
    // Enqueue the task
    [self.queue enqueueTaskWithName:@"PostThisVitalThing" userInfo:@{"wow" : "so important"}];

Also worth mentioning is the `KTBTaskAlwaysRetry` value, which will prevent a queue from abandoning a task due to too many retries:

    [self.queue enqueueTask:[KTBTask taskWithName:@"IReallyWantThisDone"
                                         userInfo:nil
                                    availableDate:nil
                                       maxRetries:KTBTaskAlwaysRetry
                                       useBackoff:YES]];

This task is available immediately (`availableDate` defaults to now) and will retry until it succeeds or the Earth is dust (or you drop your phone in the toilet). Whichever comes first.

## Requirements

ARC, iOS 6+, and [FMDB](https://github.com/ccgus/fmdb) by Gus Mueller. FMDB will be imported automatically by Cocoapods. Don't worry; it's super-small.

## Feedback

Please feel free to open issues and submit pull requests here. Thanks!

## Author

Kevin Barrett, kevin@littlespindle.com

If this helps you out or you think it's neat, check out the app that inspired it: http://airendipity.com.

## License

KTBTaskQueue is available under the MIT license. See the LICENSE file for more info.



CHANGELOG.md
# KTBTaskQueue CHANGELOG

## 0.1.0

Initial release.


Tests/KTBTaskQueueTests/main.m
//
//  main.m
//  KTBTaskQueueTests
//
//  Created by Kevin Barrett on 1/29/14.
//  Copyright (c) 2014 Little Spindle, LLC. All rights reserved.
//

#import <UIKit/UIKit.h>

#import "KTBAppDelegate.h"

int main(int argc, char * argv[])
{
    @autoreleasepool {
        return UIApplicationMain(argc, argv, nil, NSStringFromClass([KTBAppDelegate class]));
    }
}


Tests/KTBTaskQueueTestsTests/TaskSpec.m
//
//  TaskSpec.m
//  KTBTaskQueue
//
//  Created by Kevin Barrett on 1/25/14.
//  Copyright (c) 2014 Little Spindle, LLC. All rights reserved.
//

#import <Kiwi/Kiwi.h>
#import <KTBTaskQueue/KTBTask.h>

@interface KTBTask (TestAdditions)
- (void)setRetryCount:(NSUInteger)retryCount;
@end

SPEC_BEGIN(TaskSpec)

describe(@"KTBTask", ^{
    it(@"cannot be retried more than maxRetries times", ^{
        KTBTask *task = [KTBTask taskWithName:@"ATask"
                                     userInfo:nil
                                availableDate:nil
                                   maxRetries:1
                                   useBackoff:YES];
        [[theValue([task canBeRetried]) should] beYes];
        
        [task setRetryCount:1];
        [[theValue([task canBeRetried]) should] beNo];
    });
    
    it(@"has a default and allows custom available dates", ^{
        KTBTask *task = [KTBTask taskWithName:@"ATask"
                                     userInfo:nil
                                availableDate:nil
                                   maxRetries:1
                                   useBackoff:YES];
        [[theValue(-[task.availableDate timeIntervalSinceNow]) should] beLessThan:theValue(1)];
        
        NSDate *customDate = [NSDate dateWithTimeIntervalSinceReferenceDate:0];
        KTBTask *datedTask = [KTBTask taskWithName:@"DatedTask"
                                          userInfo:nil
                                     availableDate:customDate
                                        maxRetries:1
                                        useBackoff:YES];
        [[datedTask.availableDate should] equal:customDate];
    });
    
    it(@"allows task-wide turning off of retry backoff", ^{
        [KTBTask setRetryWithBackoffDefault:NO];
        KTBTask *defaultTaskNo = [KTBTask taskWithName:@"DefaultTask" userInfo:nil];
        [[theValue(defaultTaskNo.retryWithBackoff) should] beNo];
        
        [KTBTask setRetryWithBackoffDefault:YES];
        KTBTask *defaultTaskYes = [KTBTask taskWithName:@"DefaultTask" userInfo:nil];
        [[theValue(defaultTaskYes.retryWithBackoff) should] beYes];

    });
});

SPEC_END


Tests/KTBTaskQueueTests/KTBAppDelegate.h
//
//  KTBAppDelegate.h
//  KTBTaskQueueTests
//
//  Created by Kevin Barrett on 1/29/14.
//  Copyright (c) 2014 Little Spindle, LLC. All rights reserved.
//

#import <UIKit/UIKit.h>

@interface KTBAppDelegate : UIResponder <UIApplicationDelegate>

@property (strong, nonatomic) UIWindow *window;

@end


KTBTaskQueue/KTBTask.h
//
//  KTBTask.h
//  KTBTask
//
//  Created by Kevin Barrett on 1/23/14.
//  Copyright (c) 2014 Little Spindle, LLC. All rights reserved.
//

#import <Foundation/Foundation.h>

extern const NSInteger KTBTaskMaxRetriesDefault;
extern const NSInteger KTBTaskAlwaysRetry;

@class FMResultSet;

/**
 A @c KTBTask is an immutable representation of a task queued in a @c KTBTaskQueue.
 As it is immutable, changing the task returned by a delegate method or block from @c KTBTaskQueue
 will not alter the task's behavior.
 */
@interface KTBTask : NSObject
/**
 A unique number identifier for the task. Only non-nil once the task has been enqueued.
 */
@property (readonly, nonatomic, strong) NSNumber *taskID;
/**
 The task's name. Does not need to be unique.
 
 Can be used to categoryize tasks, identify individual tasks, or pretty much anything else a string can do.
 */
@property (readonly, nonatomic, copy) NSString *name;
/**
 A dictionary used to attach arbitrary data to the task. Defaults to an empty dictionary.
 
 @note Keys must be strings and values much be JSON-serializable.
 */
@property (readonly, nonatomic, strong) NSDictionary *userInfo;
/**
 Date this task was created.
 */
@property (readonly, nonatomic, strong) NSDate *createdDate;
/**
 Date after which this task is available for dequeuing and executing. Defaults to now.
 */
@property (readonly, nonatomic, strong) NSDate *availableDate;
/**
 The number of times this task has been retried.
 @note This is equivalent to the number of times this task has been attempted + 1 for the first (untried) attempt.
 */
@property (readonly, nonatomic, assign) NSInteger retryCount;
/**
 The maximum number of times this task will be retried. Defaults to 10.
 Setting this to @c KTBTaskAlwaysRetry will prevent the task from ever being abandoned due to high retry count. The task will be tried until it succeeds.
 @note The first attempt does not count as a retry, so a task will be tried a total of @c maxRetries+1 times.
 */
@property (readonly, nonatomic, assign) NSInteger maxRetries;
/**
 Whether to retry a task immediately or wait for some time before retrying. Defaults to @c YES. The default can be changed with the setRetryWithBackoffDefault: class method.
 @note When @c YES, the amount of time waited between retries increases with each retry.
 */
@property (readonly, nonatomic, assign) BOOL retryWithBackoff;

/**
 Simple constructor to return a task. The task will be available immediately, will abandon after 10 retries, and will use backoff.
 @param name The name of the task.
 @param userInfo Arbitrary JSON-encodable data for this task.
 */
+ (instancetype)taskWithName:(NSString *)name userInfo:(NSDictionary *)userInfo;
/**
 Full constructor. Passing in @c nil for @c userInfo and @c availableDate will treat them as the default: an empty dictionary and now, respectively.
 @param name The name of the task.
 @param userInfo Arbitrary JSON-encodable data for this task.
 @param availableDate Date this task should be available for execution. Passing @c nil will make the task available immediately.
 @param maxRetries Number of times to retry this task. Passing in @c KTBTaskAlwaysRetry will prevent this task from abandoning due to high retry count.
 @param useBackoff Whether to retry the task with exponential backoff delay or immediately.
 */
+ (instancetype)taskWithName:(NSString *)name userInfo:(NSDictionary *)userInfo availableDate:(NSDate *)availableDate maxRetries:(NSUInteger)maxRetries useBackoff:(BOOL)useBackoff;
/**
 Changes the default value of @c retryWithBackoff on newly-created tasks. The default is @c YES.
 */
+ (void)setRetryWithBackoffDefault:(BOOL)defaultRetryWithBackoffValue;
/**
 @return @c YES if the task can be retried.
 */
- (BOOL)canBeRetried;

// Ignore me!
+ (instancetype)taskWithResultSet:(FMResultSet *)resultSet;
@end


KTBTaskQueue/KTBTaskQueue.h
//
//  KTBTaskQueue.h
//  KTBTaskQueue
//
//  Created by Kevin Barrett on 1/23/14.
//  Copyright (c) 2014 Little Spindle, LLC. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "KTBTaskQueueDelegate.h"
#import "KTBTask.h"

typedef void(^KTBTaskQueueExecutionBlock)(KTBTask *task, KTBTaskCompletionBlock completion);

@interface KTBTaskQueue : NSObject
@property (readwrite, nonatomic, weak) id<KTBTaskQueueDelegate> delegate;
/**
 @c YES if the queue is not actively dequeuing jobs, and @c NO if it is.

 Defaults to @c NO. Setting to @c YES will suspended the queue and prevent tasks from running
 until @c suspended is set to NO again.
 
 @note @c KTBTaskQueue is initialized in a non-suspended state, so the first enqueued
 task will automatically be run. Adding a task to a suspended queue does not un-suspend it.
 Setting @c suspended to @c NO will cause any tasks added in the interim to be run.
 */
@property (readwrite, nonatomic, assign, getter = isSuspended) BOOL suspended;
/**
 This is set to @c YES when the queue is actually processing a task, a serial action. @c NO when the queue is idle.
 */
@property (readonly, atomic, assign, getter = isProcessing) BOOL processing;
/**
 If set, this block will be used instead of the @c taskQueue:executeTask:completion:
 delegate method to execute tasks.
 @note The other (optional) delegate methods will still be called, if implemented.
 */
@property (readwrite, nonatomic, copy) KTBTaskQueueExecutionBlock executionBlock;
/**
 Setting @c prohibitsBackoff to @c YES will force tasks added to the queue to retry
 immediately without respect to the individual task's @c retryWithBackoff property.
 */
@property (readwrite, nonatomic, assign) BOOL prohibitsBackoff;
/**
 The time interval in seconds used to determine backoff retry times and how often the queue
 should check for available tasks. Defaults to 10 seconds.
 @note On iOS 7+ the queue will use the @c setTolerance: method on @c NSTimer to allow the
 polling timer to be coalesced with other timers by the system (and thus save energy),
 and as such this time interval may not be strictly adhered to.
 */
@property (readwrite, nonatomic, assign) NSTimeInterval backoffPollingInterval;

/**
 A constuctor that returns a queue persisted at @c filePath. The queue will save tasks to disk
 at that location. As such, tasks will persist and be retried across app launches as long as 
 @c filePath remains the same.
 @param filePath The location on disk to store the queue.
 @return A new disk-based queue.
 */
+ (instancetype)queueAtPath:(NSString *)filePath;
/**
 A constructor that returns a queue that exists only in memory. Useful for one-off task tracking
 or testing.
 @return A new memory-based queue.
 */
+ (instancetype)queueInMemory;
/**
 Designated initializer.
 @param filePath A path on disk to store the queue, or @c nil if the queue should only exist in memory.
 @return A new queue.
 */
- (instancetype)initWithPath:(NSString *)filePath;

/**
 A shorthand way of enqueuing tasks when all you need is a name and a dictionary. Task will (almost)
 immediately be attempted via delegate or execution block.
 @param name The name of the task to enqueue.
 @param userInfo An optional dictionary of data to store in the task. Must be JSON-serializable.
 */
- (void)enqueueTaskWithName:(NSString *)name userInfo:(NSDictionary *)userInfo;
/**
 Enqueue a task. See @c KTBTask.h for how to create tasks. Task will (almost)
 immediately be attempted via delegate or execution block.
 @param task The task to enqueue.
 */
- (void)enqueueTask:(KTBTask *)task;

/**
 The number of tasks in the queue.
 */
- (NSUInteger)count;
/**
 Whether the queue contains a task with the given name.
 @param name Name of task to look for in the queue.
 */
- (BOOL)containsTaskWithName:(NSString *)name;
/**
 Whether the queue contains this exact task.
 Task must have already been enqueued and should be an argument passed to the delegate or execution block, not a new task.
 @param task Task to look for. Must be a product of the queue, not a newly created task.
 */
- (BOOL)containsTask:(KTBTask *)task;

/**
 Deletes the queue from disk or memory and renders it invalid.
 @note Tasks enqueued after the queue has been deleted are never attempted. Tasks in the queue when it is deleted are lost.
 */
- (void)deleteQueue;

@end


KTBTaskQueue/KTBTaskQueue.m
//
//  KTBTaskQueue.m
//  KTBTaskQueue
//
//  Created by Kevin Barrett on 1/23/14.
//  Copyright (c) 2014 Little Spindle, LLC. All rights reserved.
//

#import "KTBTaskQueue.h"
#import "KTBTask.h"
#import <FMDB/FMDatabaseQueue.h>
#import <FMDB/FMDatabase.h>

void KTBDispatchSyncOnMainQueue(void (^block)(void)) {
    if ([NSThread isMainThread]) {
        block();
    }
    else {
        dispatch_sync(dispatch_get_main_queue(), block);
    }
}

void KTBDispatchAsyncOnMainQueue(void (^block)(void)) {
    if ([NSThread isMainThread]) {
        block();
    }
    else {
        dispatch_async(dispatch_get_main_queue(), block);
    }
}

const NSTimeInterval KTBTaskQueueDefaultPollingInterval = 10;

@interface KTBTask (QueueAdditions)
- (NSString *)userInfoString;
- (NSNumber *)createdDateNumber;
- (NSNumber *)availableDateNumber;
- (NSDate *)nextAvailableDateWithBackoffInterval:(NSTimeInterval)backoffInterval;
@end

@interface KTBTaskQueue ()
@property (readwrite, nonatomic, strong) FMDatabaseQueue *databaseQueue;
@property (readwrite, nonatomic, copy) NSString *pathToDatabase;
@property (readwrite, nonatomic, strong) NSTimer *pollingTimer;
@property (readwrite, atomic, assign, getter = isProcessing) BOOL processing;
/**
 A valid queue has not been deleted with @c deleteQueue.
 Once a queue is deleted it is considered invalid and will no longer accept tasks.
 */
@property (readwrite, nonatomic, assign) BOOL valid;
@end

@implementation KTBTaskQueue

+ (instancetype)queueAtPath:(NSString *)filePath {
    return [[self alloc] initWithPath:filePath];
}

+ (instancetype)queueInMemory {
    return [[self alloc] initWithPath:nil];
}

- (instancetype)initWithPath:(NSString *)filePath {
    self = [super init];
    if (self) {
        if (filePath) {
            NSURL *pathURL = [NSURL fileURLWithPath:filePath];
            if (![[pathURL pathExtension] length]) {
                filePath = [filePath stringByAppendingString:@".db"];
            }
        }
        
        [self setupDatabaseQueueAtPath:filePath];
        
        _suspended = NO;
        self.processing = NO;
        _prohibitsBackoff = NO;
        self.valid = YES;
        
        _backoffPollingInterval = KTBTaskQueueDefaultPollingInterval;
        [self startPollingTimer];
    }
    return self;
}

- (instancetype)init {
    return [self initWithPath:nil];
}

- (void)dealloc {
    _valid = NO;
    
    // Stop timer and close database
    [self stopPollingTimer];
    [self.databaseQueue close];
    _databaseQueue = nil;
}

#pragma mark - Control

- (void)setSuspended:(BOOL)suspended {
    _suspended = suspended;
    
    if (suspended) {
        [self stopPollingTimer];
    }
    else {
        [self dequeueNextTask];
    }
}

- (void)setProhibitsBackoff:(BOOL)prohibitsBackoff {
    _prohibitsBackoff = prohibitsBackoff;
    
    if (prohibitsBackoff) {
        if (self.pollingTimer) {
            [self stopPollingTimer];
        }
    }
    else if (self.valid) {
        [self dequeueNextTask];
    }
}

- (void)setBackoffPollingInterval:(NSTimeInterval)backoffPollingInterval {
    _backoffPollingInterval = backoffPollingInterval;
    
    // Reset polling timer
    if (self.pollingTimer) {
        [self stopPollingTimer];
        [self dequeueNextTask];
    }
}

#pragma mark - Adding Tasks

- (void)enqueueTaskWithName:(NSString *)name userInfo:(NSDictionary *)userInfo {
    [self enqueueTask:[KTBTask taskWithName:name userInfo:userInfo]];
}

- (void)enqueueTask:(KTBTask *)task {
    if (self.valid) {
        [self insertTask:task];
        [self dequeueNextTask];
    }
}

#pragma mark - Running Tasks and Timing

- (void)startPollingTimer {
    KTBDispatchSyncOnMainQueue(^{
        if (!self.pollingTimer && !self.suspended) {
            // The polling timer repeats until one or more tasks are available for dequeueing.
            // The tasks are dequeued serially. Once all available tasks have been dequeued, the timer is resumed.
            self.pollingTimer = [NSTimer scheduledTimerWithTimeInterval:self.backoffPollingInterval target:self selector:@selector(pollingTimerDidFire:) userInfo:nil repeats:YES];
            if ([self.pollingTimer respondsToSelector:@selector(setTolerance:)]) {
                // We can add a lot of tolerance because it's not terribly important when the timer fires.
                [self.pollingTimer setTolerance:self.backoffPollingInterval / 2];
            }
        }
    });
}

- (void)stopPollingTimer {
    KTBDispatchSyncOnMainQueue(^{
        [self.pollingTimer invalidate];
        self.pollingTimer = nil;
    });
    
}

- (void)pollingTimerDidFire:(NSTimer *)timer {
    [self dequeueNextTask];
}

- (void)dequeueNextTask {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        if (self.valid && !self.suspended && !self.processing && [self hasEligibleTasks]) {
            self.processing = YES;
            KTBTask *task = [self nextTask];
            if (task) {
                // Stop the timer
                [self stopPollingTimer];
                
                // Set completion block for responding to task results
                KTBTaskCompletionBlock completionBlock = ^(KTBTaskStatus result) {
                    // Respond to whatever the was to passed us
                    [self respondToResult:result forTask:task];
                    
                    // We're no longer processing.
                    // Either there are more tasks and we should move to the next one,
                    // or we should go back to polling intermittently.
                    self.processing = NO;
                    if ([self hasEligibleTasks]) {
                        KTBDispatchAsyncOnMainQueue(^{
                            [self dequeueNextTask];
                        });
                    }
                    else {
                        [self startPollingTimer];
                    }
                };
                
                if (self.executionBlock) {
                    // Use our assigned execution block to process the task
                    self.executionBlock(task, completionBlock);
                }
                else {
                    // Ask the delegate to process the task
                    [self.delegate taskQueue:self executeTask:task completion:completionBlock];
                }
            }
            else {
                // There are no available tasks. Go back to polling.
                self.processing = NO;
                [self startPollingTimer];
            }
        }
    });
}

- (void)respondToResult:(KTBTaskStatus)result forTask:(KTBTask *)task {
    if (result == KTBTaskStatusSuccess) {
        // Remove the job from the queue
        [self deleteTask:task];
    }
    else if (result == KTBTaskStatusFailure) {
        if ([task canBeRetried]) {
            [self setRetryDataForTask:task];
        }
        else {
            [self abandonTask:task];
        }
    }
    else if (result == KTBTaskStatusAbandon) {
        [self abandonTask:task];
    }
}

- (void)abandonTask:(KTBTask *)task {
    if ([self.delegate respondsToSelector:@selector(taskQueue:willAbandonTask:)]) {
        [self.delegate taskQueue:self willAbandonTask:task];
    }
    [self deleteTask:task];
}

#pragma mark - Querying the Queue

- (NSUInteger)count {
    return [self numberOfTasks];
}

- (BOOL)containsTaskWithName:(NSString *)name {
    return [self taskWithName:name] != nil;
}

- (BOOL)containsTask:(KTBTask *)task {
    return [self taskWithID:task.taskID] != nil;
}

#pragma mark - Cleanup

// This is typically only used for testing
- (void)deleteQueue {
    // Queue is no longer valid, stop accepting and processing tasks
    self.valid = NO;
    self.processing = NO;

    // Stop timer and close database
    [self stopPollingTimer];
    [self.databaseQueue close];
    self.databaseQueue = nil;
    
    // Remove database from disk
    NSFileManager *manager = [NSFileManager new];
    [manager removeItemAtPath:self.pathToDatabase error:NULL];
    self.pathToDatabase = nil;
}

#pragma mark - Database Operations

- (void)setupDatabaseQueueAtPath:(NSString *)filePath {
    self.pathToDatabase = filePath;
    self.databaseQueue = [FMDatabaseQueue databaseQueueWithPath:filePath];
    // Create table
    [self.databaseQueue inDatabase:^(FMDatabase *db) {
        [db executeUpdate:
         @"CREATE TABLE IF NOT EXISTS tasks ("
         @"    id               INTEGER PRIMARY KEY AUTOINCREMENT,"
         @"    name             TEXT NOT NULL DEFAULT '',"
         @"    userInfo         TEXT NOT NULL DEFAULT '{}',"
         @"    createdDate      INTEGER NOT NULL,"
         @"    availableDate    INTEGER NOT NULL,"
         @"    retryCount       INTEGER NOT NULL DEFAULT 0,"
         @"    maxRetries       INTEGER NOT NULL DEFAULT 10,"
         @"    retryWithBackoff INTEGER NOT NULL DEFAULT 1"
         @");"
         ];
        [self checkErrorForDatabase:db stepDescription:[NSString stringWithFormat:@"initializing task queue at %@", filePath]];
    }];
}

- (void)insertTask:(KTBTask *)task {
    [self.databaseQueue inDatabase:^(FMDatabase *db) {
        [db executeUpdate:
         @"INSERT INTO tasks (name, userInfo, createdDate, availableDate, maxRetries, retryWithBackoff)"
         @"VALUES (?, ?, ?, ?, ?, ?)",
         task.name, [task userInfoString], [task createdDateNumber], [task availableDateNumber], @(task.maxRetries), @(!self.prohibitsBackoff && task.retryWithBackoff)];
        [self checkErrorForDatabase:db stepDescription:@"inserting task into task queue"];
    }];
}

- (BOOL)hasEligibleTasks {
    __block BOOL hasTasks = NO;
    
    [self.databaseQueue inDatabase:^(FMDatabase *db) {
        NSNumber *availableNowDateNumber = @([[NSDate date] timeIntervalSince1970]);
        FMResultSet *resultSet = [db executeQuery:@"SELECT count(id) AS count FROM tasks WHERE availableDate <= ?", availableNowDateNumber];
        [self checkErrorForDatabase:db stepDescription:@"checking for eligible tasks"];
        
        if ([resultSet next]) {
            hasTasks = ([resultSet intForColumn:@"count"] > 0);
        }
        
        [resultSet close];
    }];
    
    return hasTasks;
}

- (KTBTask *)nextTask {
    __block KTBTask *task = nil;
    [self.databaseQueue inDatabase:^(FMDatabase *db) {
        NSNumber *availableNowDateNumber = @([[NSDate date] timeIntervalSince1970]);
        FMResultSet *resultSet = [db executeQuery:@"SELECT * FROM tasks WHERE availableDate <= ? ORDER BY availableDate ASC LIMIT 1", availableNowDateNumber];
        [self checkErrorForDatabase:db stepDescription:@"selecting next eligible task"];
        
        if ([resultSet next]) {
            task = [KTBTask taskWithResultSet:resultSet];
        }
        
        [resultSet close];
    }];
    
    return task;
}

- (void)deleteTask:(KTBTask *)task {
    [self.databaseQueue inDatabase:^(FMDatabase *db) {
        [db executeUpdate:@"DELETE FROM tasks WHERE id = ?", task.taskID];
        [self checkErrorForDatabase:db stepDescription:@"deleting task from task queue"];
    }];
}

- (void)setRetryDataForTask:(KTBTask *)task {
    NSDate *nextAvailableDate = [task nextAvailableDateWithBackoffInterval:self.backoffPollingInterval];
    if ([self.delegate respondsToSelector:@selector(taskQueue:willDelayRetryOfTask:untilDate:)]) {
        nextAvailableDate = [self.delegate taskQueue:self willDelayRetryOfTask:task untilDate:nextAvailableDate];
    }
    
    [self.databaseQueue inDatabase:^(FMDatabase *db) {
        NSNumber *nextAvailableDateNumber = @([nextAvailableDate timeIntervalSince1970]);
        [db executeUpdate:@"UPDATE tasks SET retryCount = retryCount + 1, availableDate = ? WHERE id = ?", nextAvailableDateNumber, task.taskID];
        [self checkErrorForDatabase:db stepDescription:@"setting retry data for task"];
    }];
}

- (NSUInteger)numberOfTasks {
    __block NSUInteger count = 0;
    
    [self.databaseQueue inDatabase:^(FMDatabase *db) {
        FMResultSet *resultSet = [db executeQuery:@"SELECT count(id) AS count FROM tasks"];
        [self checkErrorForDatabase:db stepDescription:@"getting number of tasks"];
        
        if ([resultSet next]) {
            count = (NSUInteger)[resultSet intForColumn:@"count"];
        }
        
        [resultSet close];
    }];
    
    return count;
}

- (KTBTask *)taskWithName:(NSString *)name {
    __block KTBTask *task = nil;
    [self.databaseQueue inDatabase:^(FMDatabase *db) {
        FMResultSet *resultSet = [db executeQuery:@"SELECT * FROM tasks WHERE name = ? LIMIT 1", name];
        [self checkErrorForDatabase:db stepDescription:@"searching for a task by name"];
        
        if ([resultSet next]) {
            task = [KTBTask taskWithResultSet:resultSet];
        }
        
        [resultSet close];
    }];
    
    return task;
}

- (KTBTask *)taskWithID:(NSNumber *)taskID {
    __block KTBTask *task = nil;
    [self.databaseQueue inDatabase:^(FMDatabase *db) {
        FMResultSet *resultSet = [db executeQuery:@"SELECT * FROM tasks WHERE id = ? LIMIT 1", taskID];
        [self checkErrorForDatabase:db stepDescription:@"searching for a task by ID"];
        
        if ([resultSet next]) {
            task = [KTBTask taskWithResultSet:resultSet];
        }
        
        [resultSet close];
    }];
    
    return task;
}

- (BOOL)checkErrorForDatabase:(FMDatabase *)db stepDescription:(NSString *)description {
    BOOL hadError = [db hadError];
    
    if (hadError) {
        if (!description) {
            description = @"doing something";
        }
        
        NSLog(@"Error %@: %@", description, [db lastError]);
    }
    
    return hadError;
}

- (NSString *)description {
    NSMutableString *description = [NSMutableString new];
    [description appendFormat:@"KTBTaskQueue (%d tasks):\n", [self numberOfTasks]];
    [self.databaseQueue inDatabase:^(FMDatabase *db) {
        FMResultSet *resultSet = [db executeQuery:@"SELECT * FROM tasks ORDER BY availableDate ASC"];
        [self checkErrorForDatabase:db stepDescription:@"dumping task list"];
        
        while ([resultSet next]) {
            KTBTask *task = [KTBTask taskWithResultSet:resultSet];
            [description appendString:[task description]];
            [description appendString:@"\n"];
        }
        
        [resultSet close];
    }];
    
    return description;
}

@end

#pragma mark - KTBTask (QueueAdditions)

@implementation KTBTask (QueueAdditions)

- (NSString *)userInfoString {
    return [[NSString alloc] initWithData:[NSJSONSerialization dataWithJSONObject:self.userInfo options:0 error:NULL] encoding:NSUTF8StringEncoding];
}

- (NSNumber *)createdDateNumber {
    return @([self.createdDate timeIntervalSince1970]);
}

- (NSNumber *)availableDateNumber {
    return @([self.availableDate timeIntervalSince1970]);
}

- (NSDate *)nextAvailableDateWithBackoffInterval:(NSTimeInterval)backoffInterval {
    if (self.retryWithBackoff) {
        // 10 * 2^retryCount - 1
        // Subtract 1 so it catches next poll
        return [NSDate dateWithTimeIntervalSinceNow:(backoffInterval * pow(2, self.retryCount)) - 1];
    }
    else {
        return self.createdDate;
    }
}

@end


KTBTaskQueue/KTBTask.m
//
//  KTBTask.m
//  KTBTask
//
//  Created by Kevin Barrett on 1/23/14.
//  Copyright (c) 2014 Little Spindle, LLC. All rights reserved.
//

#import "KTBTask.h"
#import <FMDB/FMResultSet.h>

const NSInteger KTBTaskMaxRetriesDefault = 10;
const NSInteger KTBTaskAlwaysRetry = -1;
static BOOL KTBTaskRetryWithBackoffDefault = YES;

@interface KTBTask ()
@property (readwrite, nonatomic, strong) NSNumber *taskID;
@property (readwrite, nonatomic, copy) NSString *name;
@property (readwrite, nonatomic, strong) NSDictionary *userInfo;
@property (readwrite, nonatomic, strong) NSDate *createdDate;
@property (readwrite, nonatomic, strong) NSDate *availableDate;
@property (readwrite, nonatomic, assign) NSInteger retryCount;
@property (readwrite, nonatomic, assign) NSInteger maxRetries;
@property (readwrite, nonatomic, assign) BOOL retryWithBackoff;
@end

@implementation KTBTask

+ (instancetype)taskWithName:(NSString *)name userInfo:(NSDictionary *)userInfo {
    return [self taskWithName:name userInfo:userInfo availableDate:[NSDate date] maxRetries:KTBTaskMaxRetriesDefault useBackoff:KTBTaskRetryWithBackoffDefault];
}

+ (instancetype)taskWithName:(NSString *)name userInfo:(NSDictionary *)userInfo availableDate:(NSDate *)availableDate maxRetries:(NSUInteger)maxRetries useBackoff:(BOOL)useBackoff {
    KTBTask *task = [KTBTask new];
    task.name = name;
    task.userInfo = userInfo ?: @{};
    task.availableDate = availableDate ?: [NSDate date];
    task.maxRetries = maxRetries;
    task.retryWithBackoff = useBackoff;
    return task;
}

+ (instancetype)taskWithResultSet:(FMResultSet *)resultSet {
    NSDictionary *resultDictionary = [resultSet resultDictionary];
    
    KTBTask *task = [KTBTask new];
    task.taskID = resultDictionary[@"id"];
    task.name = resultDictionary[@"name"];
    task.userInfo = [NSJSONSerialization JSONObjectWithData:[resultDictionary[@"userInfo"] dataUsingEncoding:NSUTF8StringEncoding] options:0 error:NULL] ?: @{};
    task.createdDate = [NSDate dateWithTimeIntervalSince1970:[resultDictionary[@"createdDate"] integerValue]];
    task.availableDate = [NSDate dateWithTimeIntervalSince1970:[resultDictionary[@"availableDate"] integerValue]];
    task.retryCount = [resultDictionary[@"retryCount"] integerValue];
    task.maxRetries = [resultDictionary[@"maxRetries"] integerValue];
    task.retryWithBackoff = [resultDictionary[@"retryWithBackoff"] boolValue];
    return task;
}

+ (void)setRetryWithBackoffDefault:(BOOL)defaultRetryWithBackoffValue {
    KTBTaskRetryWithBackoffDefault = defaultRetryWithBackoffValue;
}

- (id)init {
    self = [super init];
    if (self) {
        self.userInfo = @{};
        self.createdDate = [NSDate date];
        self.availableDate = [NSDate date];
    }
    return self;
}

- (BOOL)canBeRetried {
    return self.maxRetries < 0 || self.retryCount < self.maxRetries;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"Task %@: %@ (%d retries, next available %@), userInfo: %@",
            self.taskID, self.name, self.retryCount, self.availableDate, self.userInfo];
}

@end
